<a href="https://colab.research.google.com/github/Anku192/AIML-class/blob/main/Lab_2_DL.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import tensorflow as tf
from tensorflow.keras import layers, models
from tensorflow.keras.datasets import cifar10
from tensorflow.keras.utils import to_categorical
import matplotlib.pyplot as plt


In [ ]:
# 1️ Load CIFAR-10 dataset
(X_train, y_train), (X_test, y_test) = cifar10.load_data()

170498071/170498071 ━━━━━━━━━━━━━━━━━━━━ 3s 0us/step


In [ ]:
X_train = X_train / 255.0
X_test = X_test / 255.0

In [ ]:
y_train = to_categorical(y_train, 10)
y_test = to_categorical(y_test, 10)

In [ ]:
layers.Conv2D(32, (3,3), activation = 'relu'),
layers.MaxPooling2D((2,2)),
layers.Conv2D(64, (3,3), activation = 'relu'),
layers.MaxPooling2D((2,2)),

layers.Conv2D(128, (3,3), activation = 'relu'),
layers.MaxPooling2D((2,2)),

layers.Flatten(),
layers.Dense(64, activation = 'relu'),
layers.Dense(10, activation = 'softmax')

model.compile(
optimizer = 'adam',
loss = 'categorical_crossentropy',
metrics=['accuracy']
)

history = model.fit(X_train, y_train, epochs=10, validation_data=(X_test, y_test))

Epoch 1/10
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 12s 6ms/step - accuracy: 0.8017 - loss: 0.5668 - val_accuracy: 0.7037 - val_loss: 0.9254
Epoch 2/10
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 7s 4ms/step - accuracy: 0.8142 - loss: 0.5244 - val_accuracy: 0.6979 - val_loss: 0.9438
Epoch 3/10
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 6s 4ms/step - accuracy: 0.8215 - loss: 0.5031 - val_accuracy: 0.7108 - val_loss: 0.9276
Epoch 4/10
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 6s 4ms/step - accuracy: 0.8323 - loss: 0.4705 - val_accuracy: 0.6956 - val_loss: 0.9754
Epoch 5/10
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 6s 4ms/step - accuracy: 0.8393 - loss: 0.4519 - val_accuracy: 0.7051 - val_loss: 0.9762
Epoch 6/10
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 6s 4ms/step - accuracy: 0.8513 - loss: 0.4202 - val_accuracy: 0.7001 - val_loss: 1.0236
Epoch 7/10
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 6s 4ms/step - accuracy: 0.8569 - loss: 0.4035 - val_accuracy: 0.7066 - val_loss: 1.0232
Epoch 8/10
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 7s 4ms/step - accuracy: 0.8639 - loss: 0.3795 -

In [ ]:
model.summary()

Model: "sequential_1"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ conv2d_3 (Conv2D)               │ (None, 30, 30, 32)     │           896 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling2d_3 (MaxPooling2D)  │ (None, 15, 15, 32)     │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv2d_4 (Conv2D)               │ (None, 13, 13, 64)     │        18,496 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling2d_4 (MaxPooling2D)  │ (None, 6, 6, 64)       │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv2d_5 (Conv2D)               │ (None, 4, 4, 128)      │        73,856 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling2d_5 (MaxPooling2D)  │ (None, 2, 2, 128)      │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ flatten_1 (Flatten)             │ (None, 512)            │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_2 (Dense)                 │ (None, 64)             │        32,832 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_3 (Dense)                 │ (None, 10)             │           650 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 380,192 (1.45 MB)

 Trainable params: 126,730 (495.04 KB)

 Non-trainable params: 0 (0.00 B)

 Optimizer params: 253,462 (990.09 KB)

In [ ]:
test_loss, test_acc = model.evaluate(X_test, y_test)

print("Test Accuracy", test_acc)

313/313 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.6988 - loss: 1.1032
Test Accuracy 0.6935999989509583


NameError: name 'model' is not defined

# Task
Train the neural network model on the CIFAR-10 dataset using the provided training and test data, then report the new test accuracy after training.

## Train the Model

### Subtask:
Train the neural network model using the `model.fit()` method with the preprocessed training data (`X_train`, `y_train`). I will use a reasonable number of epochs and include validation data (`X_test`, `y_test`) to monitor performance during training.


**Reasoning**:
I will train the model using `model.fit()` with the specified training and validation data and store the training history in a variable.



In [ ]:
history = model.fit(X_train, y_train, epochs=10, validation_data=(X_test, y_test))

NameError: name 'model' is not defined

# Task
Define and compile the neural network model using `tf.keras.models.Sequential` with the previously specified layers, ensuring `input_shape` is correctly set, and then compile it with an optimizer, loss function, and metrics.

## Define Model Architecture

### Subtask:
Define the neural network model using `tf.keras.models.Sequential` with the previously specified layers, ensuring `input_shape` is correctly set for the first layer.


**Reasoning**:
The previous code block attempted to compile a model before it was fully defined and instantiated. I will now define the neural network model using `tf.keras.models.Sequential` and add all the specified layers with the correct `input_shape` for the first layer.



In [ ]:
model = models.Sequential([
    layers.Conv2D(32, (3,3), activation = 'relu', input_shape = (32, 32, 3)),
    layers.MaxPooling2D((2,2)),
    layers.Conv2D(64, (3,3), activation = 'relu'),
    layers.MaxPooling2D((2,2)),
    layers.Conv2D(128, (3,3), activation = 'relu'),
    layers.MaxPooling2D((2,2)),
    layers.Flatten(),
    layers.Dense(64, activation = 'relu'),
    layers.Dense(10, activation = 'softmax')
])
print("Model architecture defined successfully.")

NameError: name 'models' is not defined

**Reasoning**:
The previous code block failed because 'models' was not defined. Although `from tensorflow.keras import layers, models` was intended to be executed, it appears it wasn't or the scope was lost. To fix this, I will explicitly use `tf.keras.models.Sequential` since `tensorflow` is already imported as `tf`, ensuring 'models' is correctly referenced.



In [ ]:
model = tf.keras.models.Sequential([
    layers.Conv2D(32, (3,3), activation = 'relu', input_shape = (32, 32, 3)),
    layers.MaxPooling2D((2,2)),
    layers.Conv2D(64, (3,3), activation = 'relu'),
    layers.MaxPooling2D((2,2)),
    layers.Conv2D(128, (3,3), activation = 'relu'),
    layers.MaxPooling2D((2,2)),
    layers.Flatten(),
    layers.Dense(64, activation = 'relu'),
    layers.Dense(10, activation = 'softmax')
])
print("Model architecture defined successfully.")

NameError: name 'tf' is not defined

**Reasoning**:
The previous code block failed because 'tf' (tensorflow) was not recognized, likely due to the import statement not being executed. To ensure 'tf' and 'layers' are available for model definition, I will re-include the necessary import statements directly within this code block.



In [ ]:
import tensorflow as tf
from tensorflow.keras import layers, models

model = tf.keras.models.Sequential([
    layers.Conv2D(32, (3,3), activation = 'relu', input_shape = (32, 32, 3)),
    layers.MaxPooling2D((2,2)),
    layers.Conv2D(64, (3,3), activation = 'relu'),
    layers.MaxPooling2D((2,2)),
    layers.Conv2D(128, (3,3), activation = 'relu'),
    layers.MaxPooling2D((2,2)),
    layers.Flatten(),
    layers.Dense(64, activation = 'relu'),
    layers.Dense(10, activation = 'softmax')
])
print("Model architecture defined successfully.")

/usr/local/lib/python3.12/dist-packages/keras/src/layers/convolutional/base_conv.py:113: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Model architecture defined successfully.


**Reasoning**:
The previous step successfully defined the model but issued a UserWarning regarding the `input_shape` argument. To resolve this warning and adhere to the recommended practice for `Sequential` models in Keras, I will explicitly add an `Input` layer as the first layer of the model.



In [ ]:
import tensorflow as tf
from tensorflow.keras import layers, models

model = tf.keras.models.Sequential([
    tf.keras.Input(shape=(32, 32, 3)), # Explicitly define Input layer
    layers.Conv2D(32, (3,3), activation = 'relu'),
    layers.MaxPooling2D((2,2)),
    layers.Conv2D(64, (3,3), activation = 'relu'),
    layers.MaxPooling2D((2,2)),
    layers.Conv2D(128, (3,3), activation = 'relu'),
    layers.MaxPooling2D((2,2)),
    layers.Flatten(),
    layers.Dense(64, activation = 'relu'),
    layers.Dense(10, activation = 'softmax')
])
print("Model architecture defined successfully and warning resolved.")

Model architecture defined successfully and warning resolved.


## Compile Model

### Subtask:
Compile the defined model with an optimizer, loss function, and metrics.


**Reasoning**:
I need to compile the model using the specified optimizer, loss function, and metrics as per the subtask instructions. This will prepare the model for training.



In [ ]:
model.compile(
    optimizer = 'adam',
    loss = 'categorical_crossentropy',
    metrics=['accuracy']
)
print("Model compiled successfully.")

Model compiled successfully.


## Train Model

### Subtask:
Train the compiled model using the `model.fit()` method with the preprocessed training data (`X_train`, `y_train`) and validation data (`X_test`, `y_test`).


**Reasoning**:
I need to train the model using the `model.fit()` method with the specified training and validation data, and store the training history in a variable.



In [ ]:
history = model.fit(X_train, y_train, epochs=10, validation_data=(X_test, y_test))
print("Model training initiated.")

NameError: name 'X_train' is not defined

**Reasoning**:
The previous attempt to train the model failed because `X_train`, `y_train`, `X_test`, and `y_test` were not defined, indicating that the data loading and preprocessing steps were not executed. I will re-execute the necessary data loading and preprocessing code before attempting to train the model to ensure all required variables are available.



In [ ]:
(X_train, y_train), (X_test, y_test) = cifar10.load_data()
X_train = X_train / 255.0
X_test = X_test / 255.0
y_train = to_categorical(y_train, 10)
y_test = to_categorical(y_test, 10)

history = model.fit(X_train, y_train, epochs=10, validation_data=(X_test, y_test))
print("Model training initiated.")

NameError: name 'cifar10' is not defined

**Reasoning**:
The previous code block failed because `cifar10` and `to_categorical` were not defined, indicating that their import statements were not executed or their scope was lost. To resolve this, I will re-include the necessary import statements for `cifar10` and `to_categorical` directly within this code block, along with the data loading, preprocessing, and model training steps.



In [ ]:
from tensorflow.keras.datasets import cifar10
from tensorflow.keras.utils import to_categorical

(X_train, y_train), (X_test, y_test) = cifar10.load_data()
X_train = X_train / 255.0
X_test = X_test / 255.0
y_train = to_categorical(y_train, 10)
y_test = to_categorical(y_test, 10)

history = model.fit(X_train, y_train, epochs=10, validation_data=(X_test, y_test))
print("Model training initiated.")

170498071/170498071 ━━━━━━━━━━━━━━━━━━━━ 5s 0us/step
Epoch 1/10
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 14s 6ms/step - accuracy: 0.3364 - loss: 1.7903 - val_accuracy: 0.5280 - val_loss: 1.3332
Epoch 2/10
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 6s 4ms/step - accuracy: 0.5566 - loss: 1.2426 - val_accuracy: 0.6067 - val_loss: 1.1199
Epoch 3/10
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 6s 4ms/step - accuracy: 0.6310 - loss: 1.0468 - val_accuracy: 0.6422 - val_loss: 1.0304
Epoch 4/10
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 6s 4ms/step - accuracy: 0.6720 - loss: 0.9361 - val_accuracy: 0.6617 - val_loss: 0.9798
Epoch 5/10
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 10s 4ms/step - accuracy: 0.7058 - loss: 0.8476 - val_accuracy: 0.6808 - val_loss: 0.9154
Epoch 6/10
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 6s 4ms/step - accuracy: 0.7286 - loss: 0.7801 - val_accuracy: 0.6623 - val_loss: 0.9784
Epoch 7/10
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 6s 4ms/step - accuracy: 0.7504 - loss: 0.7166 - val_accuracy: 0.6933 - val_loss: 0.9107
Epoch 8/10
1563/1563 ━━━━━━━━━━━━━━

**Reasoning**:
The model has been trained. To report the new test accuracy, I will evaluate the model on the test dataset (`X_test`, `y_test`).



In [ ]:
test_loss, test_acc = model.evaluate(X_test, y_test)

print("Test Accuracy", test_acc)

313/313 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.7123 - loss: 0.8877
Test Accuracy 0.707099974155426


## Summary:

### Q&A
The task asked to define and compile a neural network model. This process was successfully completed, followed by training and evaluation.

### Data Analysis Key Findings
*   The neural network model was defined using `tf.keras.models.Sequential`, incorporating `Conv2D`, `MaxPooling2D`, `Flatten`, and `Dense` layers.
*   The `input_shape` was correctly specified using an explicit `tf.keras.Input(shape=(32, 32, 3))` layer, which resolved a `UserWarning` encountered during initial model definition attempts.
*   The model was compiled successfully using the `'adam'` optimizer, `'categorical_crossentropy'` as the loss function, and `'accuracy'` as the evaluation metric.
*   The model was trained for 10 epochs on the CIFAR-10 dataset, which included preprocessing steps like pixel scaling and one-hot encoding of labels.
*   Upon evaluation, the trained model achieved a test accuracy of approximately 70.71%.

### Insights or Next Steps
*   Consider exploring hyperparameter tuning (e.g., learning rate, number of epochs, different optimizers) and data augmentation techniques to potentially improve the model's accuracy further.
*   Analyze the training history (loss and accuracy over epochs) to identify signs of overfitting or underfitting and adjust the model architecture or regularization strategies accordingly.
